java/ql/src/experimental/Security/CWE/CWE-918/RequestForgery.qll

```diff
@@ -167,6 +168,61 @@ private class RequestEntityUriArg extends RequestForgerySink {
   }
 }
 
+/**
+ * A class representing a sink for remote class loading mechanism.
+ */
+private class URLClassloaderSink extends RequestForgerySink {
+  URLClassloaderSink() {
+
+    exists(MethodAccess e |
+      e.getMethod().getDeclaringType().getASupertype*() instanceof TypeClassLoader and
+      e.getQualifier() = this.asExpr() and
+      e.getMethod().getName() in ["loadClass", "findResource", "findResources", "getResourceAsStream"]
+    ) 
+  }
+}
+
+/** Connect URLClassLoader instantiation to the constructing reference */
+private class TypeURLClassLoaderTaintPreservingMethod extends Constructor, TaintPreservingCallable { 
+  TypeURLClassLoaderTaintPreservingMethod() { 
+    this.getDeclaringType() instanceof TypeURLClassLoader
+  } 
+ 
+  override predicate returnsTaintFrom(int arg) { arg = 0 } 
+} 
+
+/** Connect URLClassLoader instantiation with newInstance method to the constructing reference */
+private class TypeURLClassLoaderNewInstanceTaintPreservingMethod extends TaintPreservingCallable { 
+  TypeURLClassLoaderNewInstanceTaintPreservingMethod() { 
+    this.getDeclaringType() instanceof TypeURLClassLoader and
+    this.hasName("newInstance")
+  } 
+ 
+  override predicate returnsTaintFrom(int arg) { arg = 0 } 
+} 
```

Comment:
I dont think these taint steps make sense out of this particular SSRF query. They should probably be moved to the SSRF tainttracking configuration /cc @owen-mc

Reply:
Do you mean to move these 2 classes from RequestForgery.qll to RequestForgery.ql?

Reply:
Yep, add them as AdditionalTaintSteps within the RequestForgery TaintTracking configuration

Reply:
Just to expand slightly on what @pwntester wrote, a class extending `TaintPreservingCallable` should be used for methods that should propagate taint in all configurations, as mentioned [here](https://github.com/github/codeql/blob/cd059eb965238127ca5ce196afc9f01322c36ff0/java/ql/src/semmle/code/java/dataflow/FlowSteps.qll#L92). For methods that should only propagate taint in a particular taint tracking configuration, the method should be mentioned in an override of [`isAdditionalTaintStep`](https://github.com/github/codeql/blob/2172fb6e650880d68747f9204d50e7af1fefc624/java/ql/src/semmle/code/java/dataflow/internal/tainttracking1/TaintTrackingImpl.qll#L101). This is for performance reasons, as it allows`TaintPreservingCallable` to be calculated once and then used for many different queries.